_posts/2023-03-06-opentelemetry.md

```diff
@@ -0,0 +1,245 @@
+```
+
+With the Jaeger instance running, an OTLP endpoint for getting the traces is exposed through the `http://simplest-collector:4317` service.
+
+> If using Minikube, you can use the port-forwarding to have access to the Jaeger Web UI easily from your local browser at `http://localhost:16686`. Just use the corresponding command `kubectl port-forward svc/simplest-query 16686:16686`.
+
+For more details, the official Jaeger documentation explains all the steps [here](https://www.jaegertracing.io/docs/latest/operator/).
+
+## Let's trace the HTTP bridge traffic
+
+Let's take the Strimzi HTTP bridge as an example of tracing to show how OpenTelemetry works.
+The first step is to deploy a `KafkaBridge` instance with the tracing enabled and the corresponding environment variables to configure the OTLP exporter.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaBridge
+metadata:
+  name: my-bridge
+spec:
+  replicas: 1
+  bootstrapServers: my-cluster-kafka-bootstrap:9092
+  http:
+    port: 8080
+  template:
+    bridgeContainer:
+      env:
+        - name: OTEL_SERVICE_NAME
+          value: my-otel-service
+        - name: OTEL_EXPORTER_OTLP_ENDPOINT
+          value: http://simplest-collector:4317
+  tracing:
+    type: opentelemetry
+```
+
+With the HTTP bridge up and running, we can use any HTTP client for sending and receiving messages.
+It could be a simple command line tool like `cURL`, a more UI oriented like [Postman](https://www.postman.com/) or whatever is your preferred one.
+You could also use an application you developed with your preferred language.
+For the purpose of this blog post, I used the Java HTTP client examples we have in the Strimzi repository [here](https://github.com/strimzi/client-examples).
+By getting the [deployment-tracing-opentelemetry.yaml](https://raw.githubusercontent.com/strimzi/client-examples/main/java/http/deployment-tracing-opentelemetry.yaml) YAML file you are able to deploy:
+
+* a producer Java HTTP client which sends a message every second to the Apache Kafka cluster through the Strimzi HTTP bridge.
+* a consumer Java HTTP client which polls for messages every second from the Apache Kafka cluster through the Strimzi HTTP bridge.
+
+Both applications are configured in order to trace the HTTP requests they send to the bridge.
+
+Get the above file and change the OpenTelemetry exporter configuration environment variables to match with what we have already deployed and configured in the `KafkaBridge` custom resource.
+
+```yaml
+env:
+  # ...
+  - name: OTEL_SERVICE_NAME
+    value: my-otel-service
+  - name: OTEL_EXPORTER_OTLP_ENDPOINT
+    value: http://simplest-collector:4317
+  # ...
+```
+
+Save and deploy the YAML:
+
+```shell
+kubectl apply -f deployment-tracing-opentelemetry.yaml
+```
+
+When the HTTP clients exchange the messages we can see the traces and the corresponding spans in the Jaeger Web UI.
+
+![](/assets/images/posts/2023-03-06-opentelemetry-01.png)
```

Comment:
Shouldn't the clients be visible in the traces as well?

Reply:
The clients sending and receiving are part of the above two traces. If you see at the other pictures, you can see both the producer and consumer traces have:

* one span about the HTTP client sending the request
* one span about the bridge handling the HTTP request
* one span about the Kafka client send/receive to/from the topic

Reply:
Not sure I follow. It is 3 spans. But all of them are from the same `my-otel-service`, so they all come from the bridge. The producer / consumer is not visible there.

Reply:
I set the same `my-otel-service` for the HTTP clients applications as well.
They are producing that spans under the same trace. From a UI perspective you don't have to jump across services in the drop down list.

Reply:
That seems a bit strange and the flow is not obvious from it. But I guess if that is how you want it, then fine. 🤷

Reply:
I got what you mean, going to run with different service names and catching a new screenshot.

Reply:
@scholzj just pushed changes, let me know what do you think now please.